### PR TITLE
shippingservice: add http client request span

### PR DESCRIPTION
**Pull request by styblope**
# Changes
I propose to add a dedicated http client span for the outgoing calls to `/getquote`. This change will provide tracing separation between the server part and the downstream requests client part inside shippingservice. A client span can also better match with the peer quoteservice server span. The client middleware-based instrumentation uses the available `reqwests-tracing` and `tracing-opentelementry` crates.

## Merge Requirements

For new features contributions please make sure you have completed the following
essential items:

* [x] `CHANGELOG.md` updated to document new feature additions
* [x] Appropriate documentation updates in the [docs](../docs/) folder

### CODE REVIEW

**Comment by julianocosta89** [comment]
Hello @styblope could you please sign the CLA?

**Review by julianocosta89 — CHANGES_REQUESTED** [review]
Thanks for the contribution @styblope!
A lot of new attributes and events were added automatically. Good one!

Could you also update the doc for the Shipping Service to reflect your changes?
https://github.com/open-telemetry/opentelemetry-demo/blob/main/docs/services/shippingservice.md

File: docs/services/shippingservice.md [hunk 1]
```diff
@@ -152,6 +152,21 @@ Adding a span event:
     info!("Tracking ID Created: {}", tid);
 ```
 
+### Reqwest HTTP client instrumentation
```
Comment (julianocosta89) [inline docs/services/shippingservice.md:155]: Just a nit pick 😅 

Could you move that to the following section?
https://github.com/styblope/opentelemetry-demo/blob/shippingservice-client-span/docs/services/shippingservice.md#adding-grpchttp-instrumentation

Maybe we could even split the section in 2:
`Adding gRPC instrumentation` with what is there.

and: `Adding HTTP instrumentation` with what you have added.

Comment (styblope) [reply docs/services/shippingservice.md:155]: yup makes sense, 1 sec

**Comment by julianocosta89** [comment]
Hey @GaryPWhite, as you were the main responsible for the `shippingservice` implementation, would you mind taking a look at this one?

File: src/shippingservice/src/shipping_service/quote.rs [hunk 2]
```diff
@@ -48,7 +50,10 @@ async fn request_quote(count: u32) -> Result<f64, Box<dyn std::error::Error>> {
     let mut reqbody = HashMap::new();
     reqbody.insert("numberOfItems", count);
 
-    let client = reqwest::Client::new();
+    let reqwest_client = reqwest::Client::new();
+    let client = ClientBuilder::new(reqwest_client)
```
Comment (GaryPWhite) [inline src/shippingservice/src/shipping_service/quote.rs:54]: personal style choice, I would rewrite this as just:

```rust
let client = ClientBuilder::new(reqwest::Client::new())
```

Up to you.

File: src/shippingservice/src/main.rs [hunk 3]
```diff
@@ -82,7 +91,8 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
         .await;
 
     init_logger()?;
-    init_tracer()?;
+    let tracer = init_tracer()?;
+    init_reqwest_tracing(tracer)?;
```
Comment (GaryPWhite) [inline src/shippingservice/src/main.rs:95]: same as other comment -- I think this one makes a little less sense with the `?` but still think this reads better as 

```
init_reqest_tracing(init_tracer()?)?;
```

my 2c.

**Review by GaryPWhite — APPROVED** [review]
this is really cool. I want to point out that this is a way of using even _downstream_ dependencies in a really adept way and that's a massive contribution of SME to this project. Thanks much for your input here.

left some comments but by no means blocking.

File: src/shippingservice/src/shipping_service/quote.rs (hunk 2 above)
Comment (styblope) [reply src/shippingservice/src/shipping_service/quote.rs:54]: Thanks for the kind words. I guess I wanted to be explicit and transparent with the new dependency, but I also agree with the concise Rusty syntax. Will rewrite the bits as you propose.

**Pull request merged**